# CAO points analysis in 2019 - 2021  


The Central Applications Office processes applications for undergraduate courses in Irish Higher Education Institutions (HEIs).

In [1]:
# Convenient HTTP requests.
import requests as rq

# Regular expressions.
import re

# Dates and times.
import datetime as dt

# Data frames.
import pandas as pd

# For downloading.
import urllib.request as urlrq

# For loading PDFs into pandas
import camelot

In [2]:
# Get the current date and time.
now = dt.datetime.now()

# Format as a string.
nowstr = now.strftime('%Y%m%d_%H%M%S')

<br>

## 2021 Points

Taken from [cao.ie](http://www2.cao.ie/points/l8.php)
***

In [3]:
# Fetch the CAO points URL.
resp = rq.get('http://www2.cao.ie/points/l8.php')

# Testing of we getting expected response (200 means OK).
resp

<Response [200]>

<br>

## Save original data set
***

In [4]:
# Create a file path for the original data.
pathhtml = 'data/cao2021_' + nowstr + '.html'

<br>

**Fixing encoding error**

If we inspect the page in hte browser, server says we should decode as per:  

```
Content-Type: text/html; charset=iso-8859-1
```
     
However, one line uses #x96 character which isn't defined in iso-8859-1.
Therefore to fix this issue we going to use very similar standart cp1252,
which includes #x96 character

In [5]:
# The server uses the wrong encoding, fix it.
original_encoding = resp.encoding

# Change to cp1252.
resp.encoding = 'cp1252'

In [6]:
# Save the original html file.
with open(pathhtml, 'w') as f:
    f.write(resp.text)

<br>

## Use regular expressions to select lines we want
***

In [7]:
# Compile the regular expression for matching lines.
re_course = re.compile(r'([A-Z]{2}[0-9]{3})(.*)')

<br>

### Loop through the lines of the response
***

In [9]:
# The file path for the csv file.
path2021 = 'data/cao2021_csv_' + nowstr + '.csv'

In [8]:
# Helper function to get rid of special charchters
def points_to_array(s):
    portfolio = ''
    if s[0] == '#':
        portfolio = '#'
    random = ''
    if s[-1] == '*':
        random = '*'
    points = ''
    for i in s:
        if i.isdigit():
            points = points +i
    return [point, portfolio, random]
#video 30 min Cleaning up the data

In [10]:
# Keep track of how many courses we process.
no_lines = 0

# Open the csv file for writing.
with open(path2021, 'w') as f:
    # Write a header row.
    f.write(','.join(['code', 'title', 'pointsR1', 'pointsR2']) + '\n')
    # Loop through lines of the response.
    for line in resp.iter_lines():
        # Decode the line, using the wrong encoding!
        dline = line.decode('cp1252')
        # Match only the lines representing courses.
        if re_course.fullmatch(dline):
            # Add one to the lines counter.
            no_lines = no_lines + 1
            # The course code.
            course_code = dline[:5]
            # The course title.
            course_title = dline[7:57].strip()
            # Round one points.
            course_points = re.split(' +', dline[60:])
            if len(course_points) != 2:
                course_points = course_points[:2]
            # Join the fields using a comma.
            linesplit = [course_code, course_title, course_points[0], course_points[1]]
            # Rejoin the substrings with commas in between.
            f.write(','.join(linesplit) + '\n')

# Print the total number of processed lines.
print (f'Total number of lines is {no_lines}.')

Total number of lines is 949.


<br>

**NB:** it was verified as of 03/11/2021 that there were 949 courses exactly in the CAO 2021 points list.
***

In [11]:
df2021 = pd.read_csv(path2021, encoding='cp1252')

In [12]:
df2021

code                                              title pointsR1  \
0    AL801     Software Design for Virtual Reality and Gaming      300   
1    AL802  Software Design in Artificial Intelligence for...      313   
2    AL803  Software Design for Mobile Apps and Connected ...      350   
3    AL805    Computer Engineering for Network Infrastructure      321   
4    AL810                                 Quantity Surveying      328   
..     ...                                                ...      ...   
944  WD211                                 Creative Computing      270   
945  WD212                    Recreation and Sport Management      262   
946  WD230           Mechanical and Manufacturing Engineering      230   
947  WD231                 Early Childhood Care and Education      266   
948  WD232                       Business Information Systems      261   

    pointsR2  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  
..       ...  
944      NaN  
945      NaN  
946      230  
947      NaN  
948      261  

[949 rows x 4 columns]

<br>

## 2020 Points

https://www.cao.ie/index.php?page=points&p=2020
***

In [13]:
# URL to CAO points for 2020
url2020 = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

<br>

### Save Original File
***

In [14]:
# Create a file path for the original data.
pathxlsx = 'data/cao2020_' + nowstr + '.xlsx'

# Save retrieved URL to the file
urlrq.urlretrieve(url2020, pathxlsx)

('data/cao2020_20211222_141616.xlsx',
 <http.client.HTTPMessage at 0x7f94b42c0070>)

<br>

### Load Spreadsheet using pandas
***

In [15]:
# Download and parse the excel spreadsheet.
df2020 = pd.read_excel(url2020, skiprows=10)

In [16]:
df2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

In [17]:
# Spot check a random row.
df2020.iloc[753]

CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Road Transport Technology and Management
COURSE CODE2                                                           LC286
R1 POINTS                                                                264
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      264
EOS Random *                                                             NaN
EOS Mid-point                                                            360
LEVEL                                                                      7
HEI                                         Limerick Institute of Technology
Test/Interview #                                                         NaN

In [18]:
# Spot check the last row.
df2020.iloc[-1]

CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Mechanical and Manufacturing Engineering
COURSE CODE2                                                           WD230
R1 POINTS                                                                253
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      253
EOS Random *                                                             NaN
EOS Mid-point                                                            369
LEVEL                                                                      8
HEI                                        Waterford Institute of Technology
Test/Interview #                                                         NaN

In [19]:
# Create a file path for the pandas data.
path2020 = 'data/cao2020_' + nowstr + '.csv'

# Save pandas data frame to disk.
df2020.to_csv(path2020)

<br>

## 2019 Points

https://www.cao.ie/index.php?page=points&p=2019
***

In [20]:
# Reading PDF that was dowloaded from CAO with camelot, as by default camelot only looks at the first page of the file
# here we specify that we want all pages
cao_tables = camelot.read_pdf('data/cao2019.pdf', pages='1-end')

In [21]:
# Testing that data is now loaded ( we can see there are 18 tables/pages)
cao_tables

<TableList n=18>

In [22]:
cao_tables[2].parsing_report

{'accuracy': 100.0, 'whitespace': 2.27, 'order': 1, 'page': 3}

<br>

We can export out tables to csv file now, but camelot will put each table in a diffeent file whish isn't convinient for us, so we need to have a walkaround to combine everything into one file
 ```
cao_tables.export('data/cao2019_' + nowstr + '.csv', f='csv')
```

In [23]:
# Here we using for loop to go through tables, convert them to dataframes and append everything 
# Idea for using appending method was taken from [1]
df2019 = pd.DataFrame()
i=0
for table in cao_tables:
    df = cao_tables[i].df
    i=i + 1
    df2019 = df2019.append(df)
df2019

0  \
0   Course Code INSTITUTION and COURSE   
1                                        
2                                AL801   
3                                AL802   
4                                AL803   
..                                 ...   
37                               WD200   
38                               WD210   
39                               WD211   
40                               WD212   
41                               WD230   

                                                    1    2    3  
0                                                      EOS  Mid  
1                     Athlone Institute of Technology            
2     Software Design with Virtual Reality and Gaming  304  328  
3                Software Design with Cloud Computing  301  306  
4   Software Design with Mobile Apps and Connected...  309  337  
..                                                ...  ...  ...  
37                                     Arts (options)  221  296  
38                       Software Systems Development  271  329  
39                                 Creative Computing  275  322  
40                    Recreation and Sport Management  274  311  
41           Mechanical and Manufacturing Engineering  273  348  

[966 rows x 4 columns]

In [24]:
# Create a file path for the data.
path2019 = 'data/cao2019_' + nowstr + '.csv'

# Save pandas data frame to disk.
df2019.to_csv(path2019)

<br>

## concat and join functions
***

In [25]:
# Selecting columns code and title from data frame  for 2021 and assigning it to courses2021
courses2021 = df2021[['code', 'title']]
courses2021

code                                              title
0    AL801     Software Design for Virtual Reality and Gaming
1    AL802  Software Design in Artificial Intelligence for...
2    AL803  Software Design for Mobile Apps and Connected ...
3    AL805    Computer Engineering for Network Infrastructure
4    AL810                                 Quantity Surveying
..     ...                                                ...
944  WD211                                 Creative Computing
945  WD212                    Recreation and Sport Management
946  WD230           Mechanical and Manufacturing Engineering
947  WD231                 Early Childhood Care and Education
948  WD232                       Business Information Systems

[949 rows x 2 columns]

In [26]:
# Selecting columns COURSE CODE2 and COURSE TITLE from data frame  for 2020 and assigning it to courses2020
courses2020 = df2020[['COURSE CODE2','COURSE TITLE']]
#Changing names of the columns to code and title
courses2020.columns = ['code', 'title']
courses2020

code                                              title
0     AC120                             International Business
1     AC137                                       Liberal Arts
2     AD101   First Year Art & Design (Common Entry,portfolio)
3     AD102  Graphic Design and Moving Image Design (portfo...
4     AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
1459  WD208                          Manufacturing Engineering
1460  WD210                       Software Systems Development
1461  WD211                                 Creative Computing
1462  WD212                    Recreation and Sport Management
1463  WD230           Mechanical and Manufacturing Engineering

[1464 rows x 2 columns]

In [27]:
# Combining two dataframs into one
allcourses = pd.concat([courses2021, courses2020], ignore_index=True)
allcourses

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
2408  WD208                          Manufacturing Engineering
2409  WD210                       Software Systems Development
2410  WD211                                 Creative Computing
2411  WD212                    Recreation and Sport Management
2412  WD230           Mechanical and Manufacturing Engineering

[2413 rows x 2 columns]

In [28]:
# Now, if we sort our data by 'code' we will see that there are a lof of duplicates, as most of the courses which run in 2020
# were also awailable in 2021
allcourses.sort_values('code')

code                                             title
175   AC120                            International Business
949   AC120                            International Business
950   AC137                                      Liberal Arts
176   AC137                                      Liberal Arts
951   AD101  First Year Art & Design (Common Entry,portfolio)
...     ...                                               ...
2411  WD212                   Recreation and Sport Management
2412  WD230          Mechanical and Manufacturing Engineering
946   WD230          Mechanical and Manufacturing Engineering
947   WD231                Early Childhood Care and Education
948   WD232                      Business Information Systems

[2413 rows x 2 columns]

In [29]:
# To confirm the titles of the courses were the same, which means those entries are duplicates
allcourses.loc[176]['title']

'Liberal Arts'

In [30]:
allcourses.loc[950]['title']

'Liberal Arts'

In [31]:
# Finds all extra copies of duplicated rows. There are currenntly 738 duplicates in our df
allcourses[allcourses.duplicated()]

code                                              title
949   AC120                             International Business
950   AC137                                       Liberal Arts
952   AD102  Graphic Design and Moving Image Design (portfo...
955   AD204                               Fine Art (portfolio)
956   AD211                         Fashion Design (portfolio)
...     ...                                                ...
2404  WD200                                     Arts (options)
2409  WD210                       Software Systems Development
2410  WD211                                 Creative Computing
2411  WD212                    Recreation and Sport Management
2412  WD230           Mechanical and Manufacturing Engineering

[738 rows x 2 columns]

In [32]:
# Returns a copy of the data frame with duplciates removed. There are now only 1675 rows remaining. (738+1675 = 2413 rows 
# were in total before we removed duplicates)
allcourses.drop_duplicates()

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
2403  WD195  Architectural and Building Information Modelli...
2405  WD205   Molecular Biology with Biopharmaceutical Science
2406  WD206                             Electronic Engineering
2407  WD207                             Mechanical Engineering
2408  WD208                          Manufacturing Engineering

[1675 rows x 2 columns]

In [33]:
# Finds all extra copies of duplicated rows. Now we are looking for duplicated in column 'code'
allcourses[allcourses.duplicated(subset=['code'])]

code                                              title
949   AC120                             International Business
950   AC137                                       Liberal Arts
951   AD101   First Year Art & Design (Common Entry,portfolio)
952   AD102  Graphic Design and Moving Image Design (portfo...
953   AD103  Textile & Surface Design and Jewellery & Objec...
...     ...                                                ...
2404  WD200                                     Arts (options)
2409  WD210                       Software Systems Development
2410  WD211                                 Creative Computing
2411  WD212                    Recreation and Sport Management
2412  WD230           Mechanical and Manufacturing Engineering

[896 rows x 2 columns]

In [34]:
# Returns a copy of the data frame with duplciates removed - based only on code. Index will be labeled strating form 0.., n-1
allcourses.drop_duplicates(subset=['code'], inplace=True, ignore_index=True)
allcourses

code                                              title
0     AL801     Software Design for Virtual Reality and Gaming
1     AL802  Software Design in Artificial Intelligence for...
2     AL803  Software Design for Mobile Apps and Connected ...
3     AL805    Computer Engineering for Network Infrastructure
4     AL810                                 Quantity Surveying
...     ...                                                ...
1512  WD188                                Applied Health Care
1513  WD205   Molecular Biology with Biopharmaceutical Science
1514  WD206                             Electronic Engineering
1515  WD207                             Mechanical Engineering
1516  WD208                          Manufacturing Engineering

[1517 rows x 2 columns]

#### Web resources used
[1] [www.analyticsvidhya.com](https://www.analyticsvidhya.com/blog/2020/08/how-to-extract-tabular-data-from-pdf-document-using-camelot-in-python/)